drivers/i2s/i2s_nrfx.c

```diff
@@ -0,0 +1,1336 @@
+		enum i2s_trigger_cmd ch_cmd = ch_tx->last_trigger_cmd;
+
+		if (next_buffers_needed(status)) {
+			switch (ch_cmd) {
+			case I2S_TRIGGER_DROP:
+				if (get_data_ret == 0) {
+					channel_tx_mem_clear(i2s, mem_block);
+				}
+				break;
+			case I2S_TRIGGER_STOP:
+				if (get_data_ret == 0) {
+					k_mem_slab_free(ch_tx->mem_slab,
+							(void **)&mem_block);
+				}
+				break;
+			case I2S_TRIGGER_DRAIN:
+				break;
+			default:
+				channel_change_state(ch_tx, I2S_STATE_ERROR);
+				return;
+			}
+		} else {
+			int ret = channel_change_state(ch_tx, I2S_STATE_READY);
+
+			if (ret < 0) {
+				channel_change_state(ch_tx, I2S_STATE_ERROR);
+				return;
+			}
+			if (ch_cmd == I2S_TRIGGER_DRAIN) {
+				return;
+			}
+		}
+		i2s->buffers.p_tx_buffer = NULL;
+	} else if (ch_tx->current_state == I2S_STATE_ERROR) {
+		return;
+	} else if (get_data_ret < 0) {
+		interface_error_service(i2s, "TX internal callback error");
+		channel_change_state(ch_tx, I2S_STATE_ERROR);
+		return;
+	} else if (channel_check_empty(ch_tx) == true) {
+		/*underrun error occurred*/
+		if (get_data_ret == 0) {
+			k_mem_slab_free(ch_tx->mem_slab,
+					(void **)&mem_block);
+		}
+		interface_error_service(i2s, "TX underrun error");
+		channel_change_state(ch_tx, I2S_STATE_ERROR);
+		return;
+	}
+	/* continue transmission */
+	p_new_buffers->p_tx_buffer = mem_block;
+}
+
+/* In case of constant rx transmission this callback:
+ * - stores in queue rx buffer which has just been received via I2S interface
+ * - allocates new rx buffer for next transfer purpose
+ * Status parameter informs about handler execution reason:
+ *   - if 1 - next buffer is needed
+ *   - if 0 - transfer is finishing
+ */
+static void channel_rx_callback(struct i2s_nrfx_data *i2s,
+			nrfx_i2s_buffers_t const *p_released,
+			u32_t status, nrfx_i2s_buffers_t *p_new_buffers)
+{
+	struct channel_str *ch_rx = &i2s->channel_rx;
+	int ret;
+
+	if (p_released->p_rx_buffer != NULL && next_buffers_needed(status)) {
+		/* Content of received buffer is valuable.
+		 * If 'EVENT_STOPPED' generated 'then next_buffers_needed()'
+		 * returns false - function 'channel_rx_store_data()' won't
+		 * execute because buffer didn't fill.
+		 */
+		ret = channel_rx_store_data(i2s,
+					   (u32_t **)&p_released->p_rx_buffer);
+		if (ret < 0) {
+			return;
+		}
+	}
+	if (ch_rx->current_state == I2S_STATE_STOPPING) {
+		/* finishing rx transfer caused by user trigger command*/
+		enum i2s_trigger_cmd ch_cmd = ch_rx->last_trigger_cmd;
+
+		if (next_buffers_needed(status)) {
+			switch (ch_cmd) {
+			case I2S_TRIGGER_DROP:
+				channel_rx_mem_clear(i2s);
+				break;
+			case I2S_TRIGGER_DRAIN:
+			case I2S_TRIGGER_STOP:
+				break;
+			default:
+				LOG_ERROR("RX Callback command unknown",
+					  (u32_t)ch_rx->current_state);
+				channel_change_state(ch_rx, I2S_STATE_ERROR);
+				return;
+			}
+		} else {
+			if (p_released->p_rx_buffer) {
+				ret = channel_rx_store_data(i2s,
+					(u32_t **)&p_released->p_rx_buffer);
```

Comment:
This part of code needs to be described. For instance, why is this data stored only when next buffers were not requested by the nrfx_i2s driver?